File: Patterns_of_Patterns/pop.cpp
#include <algorithm>
#include <iostream>
#include <memory>
#include <vector>


class Quackable {
   public:
    virtual void quack() = 0;
};

class AbstractDuckFactory {
   public:
    virtual std::unique_ptr<Quackable> createMallardDuck() = 0;
    virtual std::unique_ptr<Quackable> createRedheadDuck() = 0;
    virtual std::unique_ptr<Quackable> createDuckCall() = 0;
    virtual std::unique_ptr<Quackable> createRubberDuck() = 0;
};

class MallardDuck : public Quackable {
   public:
    void quack() override { std::cout << "QUACKKKKKK!!!\n"; }
};

class RedheadDuck : public Quackable {
   public:
    void quack() override { std::cout << "QUACKKKKKK!!!\n"; }
};

class DuckCall : public Quackable {
   public:
    void quack() override { std::cout << "KUAWWKKKKKK!!!\n"; }
};

class RubberDuck : public Quackable {
   public:
    void quack() override { std::cout << "Squack\n"; }
};

class Goose {
   public:
    void honk() { std::cout << "HONK!!!\n"; }
};

class GooseAdapter : public Quackable {
    std::unique_ptr<Goose> goose;

   public:
    explicit GooseAdapter(std::unique_ptr<Goose> goose)
        : goose(std::move(goose)) {}

    void quack() override { this->goose->honk(); }
};

class DuckFactory : public AbstractDuckFactory {
   public:
    std::unique_ptr<Quackable> createMallardDuck() override { return std::make_unique<MallardDuck>(); }

    std::unique_ptr<Quackable> createRedheadDuck() override { return std::make_unique<RedheadDuck>(); }

    std::unique_ptr<Quackable> createDuckCall() override { return std::make_unique<DuckCall>(); }

    std::unique_ptr<Quackable> createRubberDuck() override { return std::make_unique<RubberDuck>(); }
};

class QuackCounter : public Quackable {
    std::unique_ptr<Quackable> duck;
    static int numQuacks;

   public:
    explicit QuackCounter(std::unique_ptr<Quackable> duck)
        : duck(std::move(duck)) {}

    static int getNumQuacks() { return numQuacks; }
    static void resetNumQuacks() { numQuacks = 0; }

    void quack() override {
        this->duck->quack();
        QuackCounter::numQuacks++;
    }
};

int QuackCounter::numQuacks = 0;

class CountingDuckFactory : public AbstractDuckFactory {
   public:
    std::unique_ptr<Quackable> createMallardDuck() override { return std::make_unique<QuackCounter>(std::make_unique<MallardDuck>()); }

    std::unique_ptr<Quackable> createRedheadDuck() override { return std::make_unique<QuackCounter>(std::make_unique<RedheadDuck>()); }

    std::unique_ptr<Quackable> createDuckCall() override { return std::make_unique<QuackCounter>(std::make_unique<DuckCall>()); }

    std::unique_ptr<Quackable> createRubberDuck() override { return std::make_unique<QuackCounter>(std::make_unique<RubberDuck>()); }
};

class Flock : public Quackable {
    std::vector<std::unique_ptr<Quackable>> quackers;

   public:
    void add(std::unique_ptr<Quackable> quacker) {
        this->quackers.push_back(std::move(quacker));
    }

    void quack() override {
        for (const auto& quacker : this->quackers) {
            quacker->quack();
        }
    }
};

class DuckSimulator {
   public:
    void simulate(std::unique_ptr<AbstractDuckFactory> duckFactory) {
        std::unique_ptr<Quackable> mallardDuck = duckFactory->createMallardDuck();
        std::unique_ptr<Quackable> redheadDuck = duckFactory->createRedheadDuck();
        std::unique_ptr<Quackable> duckCall = duckFactory->createDuckCall();
        std::unique_ptr<Quackable> rubberDuck = duckFactory->createRubberDuck();
        std::unique_ptr<Quackable> gooseDuck = std::make_unique<QuackCounter>(std::make_unique<GooseAdapter>(std::make_unique<Goose>()));

        auto floackOfDucks = std::make_unique<Flock>();
        floackOfDucks->add(std::move(redheadDuck));
        floackOfDucks->add(std::move(duckCall));
        floackOfDucks->add(std::move(rubberDuck));
        floackOfDucks->add(std::move(gooseDuck));

        auto flockOfMallards = std::make_unique<Flock>();
        flockOfMallards->add(std::move(duckFactory->createMallardDuck()));
        flockOfMallards->add(std::move(duckFactory->createMallardDuck()));
        flockOfMallards->add(std::move(duckFactory->createMallardDuck()));

        floackOfDucks->add(std::move(flockOfMallards));

        std::cout << "Duck Simulator: Whole Flock Simulation\n";

        this->simulate(std::move(floackOfDucks));

        std::cout << "Toatal quacks performed: " << QuackCounter::getNumQuacks() << "\n";
    }

    void simulate(std::unique_ptr<Quackable> duck) {
        duck->quack();
    }
};

int main() {
    auto duckSimulator = std::make_unique<DuckSimulator>();
    std::unique_ptr<AbstractDuckFactory> duckfactory = std::make_unique<CountingDuckFactory>();
    duckSimulator->simulate(std::move(duckfactory));
}

File: Patterns_of_Patterns/pop.py
from abc import ABC, abstractmethod

class Quackable(ABC):
    @abstractmethod
    def quack(self) -> None: pass

class AbstractDuckFactory(ABC):
    @abstractmethod
    def createMallardDuck(self)-> Quackable: pass

    @abstractmethod
    def createRedheadDuck(self)-> Quackable: pass
    
    @abstractmethod
    def createDuckCall(self)-> Quackable: pass
    
    @abstractmethod
    def createRubberDuck(self)-> Quackable: pass

class MallardDuck(Quackable):
    def quack(self) -> None: print("QUACK!!")
    
class RedheadDuck(Quackable):
    def quack(self) -> None: print("QUACK!!")
    
class DuckCall(Quackable):
    def quack(self) -> None: print("KWAWK!!")
    
class RubberDuck(Quackable):
    def quack(self) -> None: print("Squeek!!")

class Goose:
    def honk(self)-> None: print("HONK!!")


class GooseAdapter(Quackable):
    def __init__(self, goose: Goose) -> None:
        self.goose: Goose = goose

    def quack(self) -> None: self.goose.honk()


class DuckFactory(AbstractDuckFactory):
    def createMallardDuck(self) -> Quackable: return MallardDuck()
    def createRedheadDuck(self) -> Quackable: return RedheadDuck()
    def createDuckCall(self) -> Quackable: return DuckCall()
    def createRubberDuck(self) -> Quackable: return RubberDuck()


class QuackCounter(Quackable):
    _numQuacks = 0
    
    @property
    def numQuacks(self) -> int: return type(self)._numQuacks

    @numQuacks.setter
    def numQuacks(self, val: int): type(self)._numQuacks = val
    
    def __init__(self, duck: Quackable) -> None:
        self.duck: Quackable = duck
    
    def resetNumQuacks(self) -> None:  self.numQuacks(0)

    def quack(self) -> None:
        self.duck.quack()
        self.numQuacks = self.numQuacks + 1

class CountingDuckFactory(AbstractDuckFactory):
    def createMallardDuck(self) -> Quackable: return QuackCounter(MallardDuck())
    def createRedheadDuck(self) -> Quackable: return QuackCounter(RedheadDuck())
    def createDuckCall(self) -> Quackable: return QuackCounter(DuckCall())
    def createRubberDuck(self) -> Quackable: return QuackCounter(RubberDuck())


class Flock(Quackable):
    def __init__(self) -> None:
        self.quackers: list[Quackable] = []

    def add(self, quacker: Quackable) -> None: self.quackers.append(quacker)

    def quack(self) -> None:
        for quacker in self.quackers: quacker.quack()
            
        
class DuckSimulator:
    def simulate(self, duckFactory: AbstractDuckFactory):
        mallardDuck: Quackable = duckFactory.createMallardDuck()
        redheadDuck: Quackable = duckFactory.createRedheadDuck()
        duckCall: Quackable = duckFactory.createDuckCall()
        rubberDuck: Quackable = duckFactory.createRubberDuck()
        gooseDuck: Quackable = GooseAdapter(Goose())

        floackOfDucks = Flock()
        floackOfDucks.add(redheadDuck)
        floackOfDucks.add(duckCall)
        floackOfDucks.add(rubberDuck)
        floackOfDucks.add(gooseDuck)

        floackOfMallards = Flock()
        floackOfMallards.add(mallardDuck)
        floackOfMallards.add(mallardDuck)
        floackOfMallards.add(mallardDuck)

        floackOfDucks.add(floackOfMallards)

        print("Duck Simulator: Whole Flock Simulation")
        self.simulateQuack(floackOfDucks)
        
        print("Duck Simulator: Mallard Flock Simulation")
        self.simulateQuack(floackOfMallards)

        print(f"Toatal Quacks performed: {QuackCounter(None).numQuacks}")

    def simulateQuack(self, quaker: Quackable)-> None: quaker.quack()


def main():
    duckSimulator = DuckSimulator()
    duckSimulator.simulate(CountingDuckFactory())


if __name__ == '__main__':
    main()

        
        
    
